### ISSUE TITLE
Training from scratch (a bit of guidance would be appreciated)

### ISSUE THREAD

**Issue by DiegoOrtego** [bertinetto/siamese-fc#61]
Hi Luca,

I would appreciate some advice so that I can approach your performace training from scratch.
I use Pytorch (already have a 40 fps version, including image loading) using your Matlab weights. Now I am training from scratch to see if I can replicate your performance (which I currently can't). Some comments/observations:

1) I have created my own curated dataset. For me and epoch is ~4K samples as I read a sample per video, thus I am training for 500 epochs instead of 50. The samples I load are separated 100 frames when possible (sometimes videos are shorter and I use the maximum). Do you use any data augmentation?

2) Score map after correlation: I have notice that from the beginning there are high responses in the center which makes me think why a random initialization already produces something useful. Is this something that you experienced?

3) I have a question about how to deal with class imbalance. Basically, the label is always the same (a circle of value 1 in the center and -1 in the rest). I think I understood from your code that you apply a weighting mask to -1 and 1 locations once the per-pixel loss is computed. However, once I apply the weights what happens is that the loss in areas of -1 goes completely down and it will have little impact when aggregating. I might be missing something

Thank you very much in advance! Hope you could take a minute to respond me!

**Comment by jvlmdr** [comment]
Hey,

1) About 500 epochs sounds good (where an epoch contains 1 image pair per sequence). Are you using different (i.e. random) pairs in each epoch? I believe we sampled frames in Uniform(1, 100), rather than always taking them 100 frames apart (but Luca might correct me here). The only data augmentation that I can remember was small translation of up to `0.5 * stride`. In some variants, we also converted images to greyscale (with probability `p`) instead of training a separate greyscale network.

2) Yes, we did notice that random initialization already yields something close to a decent result. I believe the performance still increased rapidly in the first few epochs, but much more slowly after this.

3) Yep, we waited the losses such that the total weight of the positives was equal to the total weight of the negatives. We didn't notice the problem that you describe. Do you definitely have no padding in your conv/pool layers? This could make the classification of negatives trivially easy for the network.

Cheers!

**Comment by DiegoOrtego** [comment]
Many thanks Jack!

1. Yes, I randomly pick a video frame as exemplar and then randomly select a separated frame for the area search. It seems a better practice not to be constrained to 100, I will change this and add a bit of augmentation. One question, with stride do you refer to the stride of the correlation operation between the features of exemplar and search area? (I am using stride 1 for all conv layers). Another detail, I am using batchsize 8.

2. Great, fast learning at the beginning is what I have experienced. However, I am currently training without balancing classes. This links with item 3.

3. I have no padding in the network to prevent what you are commenting. However, I am not currently dealing with class imbalance as I am not sure if I am doing it properlly (I include key parts of my code).

Thinking about label as 2d array with 1 and -1 (radius=16) I have replicated your code:
    sumP = np.sum(label == 1)
    sumN = np.sum(label == -1)
    instanceWeight = np.ones(label.shape, dtype='float32')
    instanceWeight[label == 1] = 0.5 * instanceWeight[label == 1] / sumP
    instanceWeight[label == -1] = 0.5 * instanceWeight[label == -1] / sumN

Then, I apply this weight to the loss (I force the loss function to return per-pixel values instead of aggregated ones) (I don't know if it is important that the loss turns very small compared to the one with no weighting):

scores = siam(z_img, x_img) ## Forward
scores_up = F.upsample(scores, size=(final_score_sz, final_score_sz), scale_factor=None, align_corners=True, mode='bilinear') ## Upsampling
score_sz = scores_up.size()
loss_t = criterion(scores_up, label) #Return loss at pixel level
loss_w = (loss_t*weight2d) #Weighting (weight2d has been expanded depending on the batch size)
#Next three lines are to aggregated while handling variable batch size
size_loss = loss_w.size()
loss_t2 = loss_w.view(size_loss[0], size_loss[1], -1)
loss = loss_t2.sum(dim=2).mean() #sum all pixels for the 2d loss of each sample + mean of all samples in the batch

Do you see something wrong?

Thank you in advance!

**Comment by jvlmdr** [comment]
Hey,

Actually, if you are using `stride = 1` in all conv and pool layers (and no dilations either), I think this could be the bigger issue. It is important to have some stride (or dilation) to increase the size of the receptive field.

For the cross-correlation between template features and search-image features, we always use `stride = 1`. There would not be much point of reducing the resolution here, it would only make your predictions coarser.

**Comment by DiegoOrtego** [comment]
Sorry, I wasn't precise. I use stride 2 in the first conv and then stride 1 for the rest. But, I also use max pooling after 1st (kernelsize 3 and stride 2) and 2nd (kernelsize 3 and stride 1) conv layers. I have replicated all the following behaviour from your tensorflow code:

_conv_stride = np.array([2,1,1,1,1])
_filtergroup_yn = np.array([0,1,0,1,1], dtype=bool)
_bnorm_yn = np.array([1,1,1,1,0], dtype=bool)
_relu_yn = np.array([1,1,1,1,0], dtype=bool)
_pool_stride = np.array([2,1,0,0,0]) # 0 means no pool
_pool_sz = 3
_bnorm_adjust = True

One final doubt. When defining the label (with 1 and -1) which is the meaning of the resp_stride parameter involved (the stride of the correlation?). I mean here:

function [net, derOutputs, inputs_fn] = setup_loss(net, resp_sz, resp_stride, loss_opts)
...
[pos_eltwise, instanceWeight] = create_labels(...
        resp_sz, loss_opts.labelWeight, ...
        loss_opts.rPos/resp_stride, loss_opts.rNeg/resp_stride);

I am currently using 1 for that value.

Best,
Diego.
